clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java

```diff
@@ -25,25 +25,40 @@
 import java.util.List;
 
 public class MetadataRequest extends AbstractRequest {
-    
+
     private static final Schema CURRENT_SCHEMA = ProtoUtils.currentRequestSchema(ApiKeys.METADATA.id);
     private static final String TOPICS_KEY_NAME = "topics";
 
     private final List<String> topics;
 
+    /**
+     * Constructor to support asking for no topic metadata
+     */
+    public MetadataRequest() {
```

Comment:
I also like the builder pattern. It reads nicely and you don't have to care about argument order. On the other hand, it's also easier to forget necessary arguments and it feels kind of silly when the number of arguments is small. One nice thing about factory methods is that you can give them convenient names (e.g. MetadataRequest.allTopicsV1()). Both options are probably better than using the constructors, but I don't think it would be too bad to stick with the current convention for this patch. 

Another option would be to include an explicit flag in the constructor. For example:
```java
public MetadataRequest(boolean includeAllTopics);
```
We could then have the other constructor which accepts the topic list use the "empty means empty" semantics, and users would have to call this method to get all topics.